modin/core/dataframe/pandas/metadata/dtypes.py

```diff
@@ -12,13 +12,524 @@
+
+if TYPE_CHECKING:
+    from modin.core.dataframe.pandas.dataframe.dataframe import PandasDataframe
+    from .index import ModinIndex
 
 from modin.error_message import ErrorMessage
 
 
+class DtypesDescriptor:
+    """
+    Describes partial dtypes.
+
+    Parameters
+    ----------
+    known_dtypes : dict[IndexLabel, np.dtype], optional
+        Columns that we know dtypes for.
+    cols_with_unknown_dtypes : list[IndexLabel], optional
+        Column names that have unknown dtypes. If specified together with `remaining_dtype`, must describe all
+        columns with unknown dtypes, otherwise, the missing columns will be assigned to `remaining_dtype`.
+        If `cols_with_unknown_dtypes` is incomplete, you must specify `know_all_names=False`.
+    remaining_dtype : np.dtype, optional
+        Dtype for columns that are not present neither in `known_dtypes` nor in `cols_with_unknown_dtypes`.
+        This parameter is intended to describe columns that we known dtypes for, but don't know their
+        names yet. Note, that this parameter DOESN'T describe dtypes for columns from `cols_with_unknown_dtypes`.
+    parent_df : PandasDataframe, optional
+        Dataframe object for which we describe dtypes. This dataframe will be used to compute
+        missing dtypes on ``.materialize()``.
+    columns_order : dict[int, IndexLabel], optional
+        Order of columns in the dataframe. If specified, must describe all the columns of the dataframe.
+    know_all_names : bool, default: True
+        Whether `known_dtypes` and `cols_with_unknown_dtypes` contain all column names for this dataframe besides those,
+        that are being described by `remaining_dtype`.
+        One can't pass `know_all_names=False` together with `remaining_dtype` as this creates ambiguity
+        on how to interpret missing columns (whether they belong to `remaining_dtype` or not).
+    _schema_is_known : bool, optional
+        Whether `known_dtypes` describe all columns in the dataframe. This parameter intended mostly
+        for internal use.
+    """
+
+    def __init__(
+        self,
+        known_dtypes: Optional[dict[IndexLabel, np.dtype]] = None,
+        cols_with_unknown_dtypes: Optional[list[IndexLabel]] = None,
+        remaining_dtype: Optional[np.dtype] = None,
+        parent_df: Optional["PandasDataframe"] = None,
+        columns_order: Optional[dict[int, IndexLabel]] = None,
+        know_all_names: bool = True,
+        _schema_is_known: Optional[bool] = None,
+    ):
+        if not know_all_names and remaining_dtype is not None:
+            raise RuntimeError(
+                "It's not allowed to pass 'remaining_dtype' and 'know_all_names=False' at the same time."
+            )
+        # columns with known dtypes
+        self._known_dtypes: dict[IndexLabel, np.dtype] = (
+            {} if known_dtypes is None else dict(known_dtypes)
+        )
+        if known_dtypes is not None and len(self._known_dtypes) != len(known_dtypes):
```

Comment:
Is this possible in the case when `known_dtypes` is initially dict?

Reply:
no it's not, this check here just in case someone will pass `pandas.Series` with duplicated indices